File: core/src/main/java/vn/loitp/views/navigationview/LTextNavigationView.java
package vn.loitp.views.navigationview;

import android.content.Context;
import android.graphics.Color;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.View;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.daimajia.androidanimations.library.Techniques;

import java.util.List;

import loitp.core.R;
import vn.loitp.core.utilities.LAnimationUtil;
import vn.loitp.core.utilities.LUIUtil;

/**
 * Created by www.muathu@gmail.com on 5/13/2017.
 */

public class LTextNavigationView extends RelativeLayout implements View.OnClickListener {
    private final String TAG = getClass().getSimpleName();
    private TextView tvPrev;
    private TextView tvNext;
    private TextView tv;
    private List<String> stringList;
    private int currenIndex;
    private boolean isEnableAnimation = true;
    private int colorOn = Color.BLACK;
    private int colorOff = Color.GRAY;
    private NVCallback nvCallback;

    public interface NVCallback {
        public void onIndexChange(int index, String s);
    }

    public void setNVCallback(NVCallback nvCallback) {
        this.nvCallback = nvCallback;
    }

    public LTextNavigationView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public LTextNavigationView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init();
    }

    private void init() {
        inflate(getContext(), R.layout.view_navigation_text, this);
        this.tv = (TextView) findViewById(R.id.tv);
        this.tvPrev = (TextView) findViewById(R.id.tv_prev);
        this.tvNext = (TextView) findViewById(R.id.tv_next);
        tvPrev.setOnClickListener(this);
        tvNext.setOnClickListener(this);
        setEnableController(tvPrev, false);
        setEnableController(tvNext, false);
    }

    private void setEnableController(TextView textView, boolean isEnable) {
        if (isEnable) {
            textView.setEnabled(true);
            textView.setClickable(true);
            textView.setTextColor(colorOn);
        } else {
            textView.setEnabled(false);
            textView.setClickable(false);
            textView.setTextColor(colorOff);
        }
    }

    public void setStringList(List<String> stringList) {
        if (stringList == null || stringList.isEmpty()) {
            return;
        }
        currenIndex = 0;
        this.stringList = stringList;
        updateUI();
    }

    public void setCurrenIndex(int index) {
        if (stringList == null || stringList.isEmpty()) {
            return;
        }
        if (index < 0 || index > stringList.size() - 1) {
            return;
        }
        currenIndex = index;
        updateUI();
    }

    private void updateUI() {
        String s = stringList.get(currenIndex);
        tv.setText(s);
        if (isEnableAnimation) {
            LAnimationUtil.play(tv, Techniques.FlipInX);
        }
        int size = stringList.size();
        if (size == 1) {
            setEnableController(tvPrev, false);
            setEnableController(tvNext, false);
        } else {
            if (currenIndex <= 0) {
                setEnableController(tvPrev, false);
                setEnableController(tvNext, true);
            } else if (currenIndex >= size - 1) {
                setEnableController(tvPrev, true);
                setEnableController(tvNext, false);
            } else {
                setEnableController(tvPrev, true);
                setEnableController(tvNext, true);
            }
        }
        if (nvCallback != null) {
            nvCallback.onIndexChange(currenIndex, s);
        }
    }

    @Override
    public void onClick(View view) {
        if (view == tvPrev) {
            if (isEnableAnimation) {
                LAnimationUtil.play(view, Techniques.Pulse);
            }
            currenIndex--;
            updateUI();
        } else if (view == tvNext) {
            if (isEnableAnimation) {
                LAnimationUtil.play(view, Techniques.Pulse);
            }
            currenIndex++;
            updateUI();
        }
    }

    public TextView getTvPrev() {
        return tvPrev;
    }

    public TextView getTvNext() {
        return tvNext;
    }

    public TextView getTv() {
        return tv;
    }

    public List<String> getStringList() {
        return stringList;
    }

    public int getCurrenIndex() {
        return currenIndex;
    }

    public int getColorOn() {
        return colorOn;
    }

    public void setColorOn(int colorOn) {
        this.colorOn = colorOn;
    }

    public int getColorOff() {
        return colorOff;
    }

    public void setColorOff(int colorOff) {
        this.colorOff = colorOff;
    }

    public boolean isEnableAnimation() {
        return isEnableAnimation;
    }

    public void setEnableAnimation(boolean enableAnimation) {
        isEnableAnimation = enableAnimation;
    }

    public void setTextPrev(String prev) {
        tvPrev.setText(prev);
    }

    public void setTextNext(String next) {
        tvPrev.setText(next);
    }

    public void setTextSize(int dpPrev, int dpText, int dpNext) {
        LUIUtil.setTextSize(tvPrev, TypedValue.COMPLEX_UNIT_DIP, dpPrev);
        LUIUtil.setTextSize(tv, TypedValue.COMPLEX_UNIT_DIP, dpText);
        LUIUtil.setTextSize(tvNext, TypedValue.COMPLEX_UNIT_DIP, dpNext);
    }
}

File: core/src/main/java/com/views/navigationview/LTextNavigationView.kt
package com.views.navigationview

import android.content.Context
import android.graphics.Color
import android.util.AttributeSet
import android.view.View
import android.widget.RelativeLayout
import android.widget.TextView
import com.R
import com.core.utilities.LAnimationUtil
import com.core.utilities.LUIUtil
import com.daimajia.androidanimations.library.Techniques

class LTextNavigationView : RelativeLayout, View.OnClickListener {
    var tvPrev: TextView? = null
    var tvNext: TextView? = null
    var tv: TextView? = null

    private var stringList = ArrayList<String>()
    private var currentIndex = 0
    var isEnableAnimation = true
    var colorOn = Color.BLACK
    var colorOff = Color.GRAY
    private var nvCallback: NVCallback? = null

    interface NVCallback {
        fun onIndexChange(index: Int, s: String?)
    }

    fun setNVCallback(nvCallback: NVCallback?) {
        this.nvCallback = nvCallback
    }

    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs) {
        init()
    }

    constructor(context: Context?, attrs: AttributeSet?, defStyle: Int) : super(
        context,
        attrs,
        defStyle
    ) {
        init()
    }

    private fun init() {
        View.inflate(context, R.layout.view_navigation_text, this)
        tv = findViewById(R.id.textView)
        tvPrev = findViewById(R.id.tvPrev)
        tvNext = findViewById(R.id.tvNext)

        tvPrev?.setOnClickListener(this)
        tvNext?.setOnClickListener(this)
        setEnableController(textView = tvPrev, isEnable = false)
        setEnableController(textView = tvNext, isEnable = false)
    }

    private fun setEnableController(textView: TextView?, isEnable: Boolean) {
        if (isEnable) {
            textView?.let {
                it.isEnabled = true
                it.isClickable = true
                it.setTextColor(colorOn)
            }
        } else {
            textView?.let {
                it.isEnabled = false
                it.isClickable = false
                it.setTextColor(colorOff)
            }
        }
    }

    fun setStringList(stringList: ArrayList<String>?) {
        if (stringList.isNullOrEmpty()) {
            return
        }
        currentIndex = 0
        this.stringList = stringList
        updateUI()
    }

    fun setCurrentIndex(index: Int) {
        if (stringList.isEmpty()) {
            return
        }
        if (index < 0 || index > stringList.size - 1) {
            return
        }
        currentIndex = index
        updateUI()
    }

    private fun updateUI() {
        val s = stringList[currentIndex]
        tv?.text = s
        if (isEnableAnimation) {
            LAnimationUtil.play(view = tv, techniques = Techniques.FlipInX)
        }
        val size = stringList.size
        if (size == 1) {
            setEnableController(textView = tvPrev, isEnable = false)
            setEnableController(textView = tvNext, isEnable = false)
        } else {
            when {
                currentIndex <= 0 -> {
                    setEnableController(textView = tvPrev, isEnable = false)
                    setEnableController(textView = tvNext, isEnable = true)
                }
                currentIndex >= size - 1 -> {
                    setEnableController(textView = tvPrev, isEnable = true)
                    setEnableController(textView = tvNext, isEnable = false)
                }
                else -> {
                    setEnableController(textView = tvPrev, isEnable = true)
                    setEnableController(textView = tvNext, isEnable = true)
                }
            }
        }
        nvCallback?.onIndexChange(index = currentIndex, s = s)
    }

    override fun onClick(view: View) {
        if (view === tvPrev) {
            if (isEnableAnimation) {
                LAnimationUtil.play(view = view, techniques = Techniques.Pulse)
            }
            currentIndex--
            updateUI()
        } else if (view === tvNext) {
            if (isEnableAnimation) {
                LAnimationUtil.play(view = view, techniques = Techniques.Pulse)
            }
            currentIndex++
            updateUI()
        }
    }

    fun getStringList(): ArrayList<String> {
        return stringList
    }

    fun getCurrentIndex(): Int {
        return currentIndex
    }

    fun setTextPrev(prev: String?) {
        tvPrev?.text = prev
    }

    fun setTextNext(next: String?) {
        tvNext?.text = next
    }

    fun setTextSize(dpPrev: Float, dpText: Float, dpNext: Float) {
        LUIUtil.setTextSize(textView = tvPrev, size = dpPrev)
        LUIUtil.setTextSize(textView = tv, size = dpText)
        LUIUtil.setTextSize(textView = tvNext, size = dpNext)
    }
}
